src/Arcus.WebApi.Security/Authentication/CertificateAuthenticationValidator.cs

```diff
@@ -0,0 +1,133 @@
+        /// </summary>
+        /// <param name="clientCertificate">The client certificate to validate.</param>
+        /// <param name="services">The collection of registered services, (ex. from the request pipeline).</param>
+        /// <returns>
+        ///     <c>true</c> when the specified <paramref name="clientCertificate"/> is valid, <c>false</c> otherwise.
+        /// </returns>
+        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="clientCertificate"/> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="services"/> is <c>null</c>.</exception>
+        internal async Task<bool> ValidateCertificate(X509Certificate2 clientCertificate, IServiceProvider services)
+        {
+            Guard.NotNull(clientCertificate, nameof(clientCertificate), "Certificate authentication validation requires a client certificate");
+            Guard.NotNull(services, nameof(services), "Certificate authentication validation requires a service object to retrieve registered services");
+
+            ILogger logger = GetLoggerOrDefault(services);
+
+            IDictionary<X509ValidationRequirement, ExpectedCertificateValue> expectedValuesByRequirement =
+                await _certificateAuthenticationConfig.GetAllExpectedCertificateValues(services, logger);
+
+            return expectedValuesByRequirement.All(keyValue =>
+            {
+                switch (keyValue.Key)
+                {
+                    case X509ValidationRequirement.SubjectName:
+                        return IsCertificateSubjectNameAllowed(clientCertificate, keyValue.Value, logger);
+                    case X509ValidationRequirement.IssuerName:
+                        return IsCertificateIssuerNameAllowed(clientCertificate, keyValue.Value, logger);
+                    case X509ValidationRequirement.Thumbprint:
+                        return IsCertificateThumbprintAllowed(clientCertificate, keyValue.Value, logger);
+                    default:
+                        throw new ArgumentOutOfRangeException(nameof(keyValue.Key), keyValue.Key, "Unknown validation type specified");
+                }
+            });
+        }
+
+        private static ILogger GetLoggerOrDefault(IServiceProvider services)
+        {
+            ILogger logger = 
+                services.GetService<ILoggerFactory>()
+                        ?.CreateLogger<CertificateAuthenticationFilter>();
+
+            if (logger != null)
+            {
+                return logger;
+            }
+
+            return NullLogger.Instance;
+        }
+
+        private static bool IsCertificateSubjectNameAllowed(X509Certificate2 clientCertificate, ExpectedCertificateValue expected, ILogger logger)
+        {
+            IEnumerable<string> certificateSubjectNames =
+                (clientCertificate.Subject ?? String.Empty)
```

Comment:
If the subject is empty, do we even need to go through the whole process?

Maybe an early return would be better

Reply:
Nah it's ok.